docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1047 @@
+provider types to support GitOps workflows excluding `clusterctl`.
+
+#### New API Types
+
+These are the new API types being defined.
+
+There are separate types for each provider type - Core, Bootstrap,
+ControlPlane, and Infrastructure. However, since each type is similar, their
+Spec and Status uses the shared types - `ProviderSpec`,  `ProviderStatus`
+respectively.
+
+If you prefer to see how the API can be used instead of reading the type
+definition feel free to jump to the [Example API Usage
+section](#example-api-usage)
+
+```golang
+// CoreProvider is the Schema for the CoreProviders API
+type CoreProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   ProviderSpec   `json:"spec,omitempty"`
+   Status ProviderStatus `json:"status,omitempty"`
+}
+
+// BootstrapProvider is the Schema for the BootstrapProviders API
+type BootstrapProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   ProviderSpec   `json:"spec,omitempty"`
+   Status ProviderStatus `json:"status,omitempty"`
+}
+
+// ControlPlaneProvider is the Schema for the ControlPlaneProviders API
+type ControlPlaneProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   ProviderSpec   `json:"spec,omitempty"`
+   Status ProviderStatus `json:"status,omitempty"`
+}
+
+// InfrastructureProvider is the Schema for the InfrastructureProviders API
+type InfrastructureProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   ProviderSpec   `json:"spec,omitempty"`
+   Status ProviderStatus `json:"status,omitempty"`
+}
+```
+
+Below you can find details about `ProviderSpec`,  `ProviderStatus`, which is
+shared among all the provider types - Core, Bootstrap, ControlPlane, and
+Infrastructure.
+
+```golang
+// ProviderSpec defines the desired state of the Provider.
+type ProviderSpec struct {
+   // Version indicates the provider version.
+   // +optional
+   Version *string `json:"version,omitempty"`
+
+   // Manager defines the properties that can be enabled on the controller manager for the provider.
+   // +optional
+   Manager ManagerSpec `json:"manager,omitempty"`
+
+   // Deployment defines the properties that can be enabled on the deployment for the provider.
+   // +optional
+   Deployment *DeploymentSpec `json:"deployment,omitempty"`
+
+   // SecretName is the name of the Secret providing the configuration
+   // variables for the current provider instance, like e.g. credentials.
+   // Such configurations will be used when creating or upgrading provider components.
+   // The contents of the secret will be treated as immutable. If changes need
+   // to be made, a new object can be created and the name should be updated.
+   // The contents should be in the form of key:value.
+   // +optional
+   SecretName *string
```

Comment:
Just adding this note. We are currently discussing the need for this `SecretName` field. 

The reason we have a secret is so that we can store credentials that may be required by the providers as part of the InfrastructureProvider setup. For example, `AWS_B64ENCODED_CREDENTIALS` or `AZURE_CLIENT_SECRET_B64`. However, if we remove the need for such sensitive information as part of declaring an `InfrastructureProvider`, then all we have left are variables like `AWS_REGION`  or `AZURE_ENVIRONMENT`, which are not considered sensitive.

So we can instead replace the `SecretName` to something like `Variables map[string][string]` and these variables can be used to evaluate the `infrastructure-components.yaml` for the specified Infra provider.

There are on-going discussions with the infra providers regarding this change and the need to have a separate entity to manage/retrieve the identity required by the management providers to manage workload clusters.

See this proposal: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/proposal/20200506-single-controller-multitenancy.md

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/3881#issuecomment-743370519

Reply:
Happy New Year Ping!
Just wanted to raise this concern and involve provider authors to provide perspective on managing credentials for the CAPI components.

Reply:
@yastij The scope of the credentials usage will be the same as it is today (pre-operator). That is, it seems for CAPA it is used to just bootstrap an IAM instance profile.

Ok...so it seems for now we'll leave the `SecretName` as is. 
This will decouple the single-controller-multitenancy work and associated credential management from this CAEP. Also it won't feel like we are rushing the credential management design in order to fulfill this CAEP.

WDYT?

Reply:
+1